# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: dom/fs/child/FileSystemRequestHandler.cpp
# Commit: a1c3f1590850
# Full Hash: a1c3f15908502273c13d28e50bc0448157f2d6f9
# Author: Jan Varga <jvarga@mozilla.com>
# Date: 2022-10-20 21:51:26
# Regressor Bug: 1792245
# File Overlap Count: 2
# Description:
#   Bug 1792245 - Improve FileSystemManager shutdown; r=dom-storage-reviewers,jesup
#   
#   Changes:
#   - new MozPromise requests are not created after shutdown
#   - existing MozPromise requests are disconnected during shutdown
# ==============================================================================

diff -r 214c4d6bf7f9 -r a1c3f1590850 dom/fs/child/FileSystemRequestHandler.cpp
--- a/dom/fs/child/FileSystemRequestHandler.cpp	Thu Oct 20 04:21:30 2022 +0000
+++ b/dom/fs/child/FileSystemRequestHandler.cpp	Thu Oct 20 04:21:30 2022 +0000
@@ -312,6 +312,11 @@
   MOZ_ASSERT(aManager);
   MOZ_ASSERT(aPromise);
 
+  if (aManager->IsShutdown()) {
+    aError.Throw(NS_ERROR_ILLEGAL_DURING_SHUTDOWN);
+    return;
+  }
+
   aManager->BeginRequest(
       [onResolve = SelectResolveCallback<FileSystemGetHandleResponse,
                                          RefPtr<FileSystemDirectoryHandle>>(
@@ -334,6 +339,11 @@
   MOZ_ASSERT(aPromise);
   LOG(("getDirectoryHandle"));
 
+  if (aManager->IsShutdown()) {
+    aError.Throw(NS_ERROR_ILLEGAL_DURING_SHUTDOWN);
+    return;
+  }
+
   if (!IsValidName(aDirectory.childName())) {
     aPromise->MaybeRejectWithTypeError("Invalid directory name");
     return;
@@ -363,6 +373,11 @@
   MOZ_ASSERT(aPromise);
   LOG(("getFileHandle"));
 
+  if (aManager->IsShutdown()) {
+    aError.Throw(NS_ERROR_ILLEGAL_DURING_SHUTDOWN);
+    return;
+  }
+
   if (!IsValidName(aFile.childName())) {
     aPromise->MaybeRejectWithTypeError("Invalid filename");
     return;
@@ -388,6 +403,11 @@
   MOZ_ASSERT(aPromise);
   LOG(("getAccessHandle"));
 
+  if (aManager->IsShutdown()) {
+    aError.Throw(NS_ERROR_ILLEGAL_DURING_SHUTDOWN);
+    return;
+  }
+
   aManager->BeginRequest(
       [request = FileSystemGetAccessHandleRequest(aFile.entryId()),
        onResolve = SelectResolveCallback<FileSystemGetAccessHandleResponse,
@@ -410,6 +430,11 @@
   MOZ_ASSERT(!aFile.entryId().IsEmpty());
   MOZ_ASSERT(aPromise);
 
+  if (aManager->IsShutdown()) {
+    aError.Throw(NS_ERROR_ILLEGAL_DURING_SHUTDOWN);
+    return;
+  }
+
   aManager->BeginRequest(
       [request = FileSystemGetFileRequest(aFile.entryId()),
        onResolve =
@@ -432,6 +457,11 @@
   MOZ_ASSERT(!aDirectory.IsEmpty());
   MOZ_ASSERT(aPromise);
 
+  if (aManager->IsShutdown()) {
+    aError.Throw(NS_ERROR_ILLEGAL_DURING_SHUTDOWN);
+    return;
+  }
+
   aManager->BeginRequest(
       [request = FileSystemGetEntriesRequest(aDirectory, aPage),
        onResolve = SelectResolveCallback<FileSystemGetEntriesResponse, bool>(
@@ -455,6 +485,11 @@
   MOZ_ASSERT(aPromise);
   LOG(("removeEntry"));
 
+  if (aManager->IsShutdown()) {
+    aError.Throw(NS_ERROR_ILLEGAL_DURING_SHUTDOWN);
+    return;
+  }
+
   if (!IsValidName(aEntry.childName())) {
     aPromise->MaybeRejectWithTypeError("Invalid name");
     return;
@@ -482,6 +517,11 @@
   MOZ_ASSERT(aPromise);
   LOG(("MoveEntry"));
 
+  if (aManager->IsShutdown()) {
+    aError.Throw(NS_ERROR_ILLEGAL_DURING_SHUTDOWN);
+    return;
+  }
+
   // reject invalid names: empty, path separators, current & parent directories
   if (!IsValidName(aNewEntry.childName())) {
     aPromise->MaybeRejectWithTypeError("Invalid name");
@@ -510,6 +550,11 @@
   MOZ_ASSERT(aPromise);
   LOG(("RenameEntry"));
 
+  if (aManager->IsShutdown()) {
+    aError.Throw(NS_ERROR_ILLEGAL_DURING_SHUTDOWN);
+    return;
+  }
+
   // reject invalid names: empty, path separators, current & parent directories
   if (!IsValidName(aName)) {
     aPromise->MaybeRejectWithTypeError("Invalid name");
@@ -539,6 +584,11 @@
   MOZ_ASSERT(!aEndpoints.childId().IsEmpty());
   MOZ_ASSERT(aPromise);
 
+  if (aManager->IsShutdown()) {
+    aError.Throw(NS_ERROR_ILLEGAL_DURING_SHUTDOWN);
+    return;
+  }
+
   aManager->BeginRequest(
       [request = FileSystemResolveRequest(aEndpoints),
        onResolve =